src/drivers/net/usbnet.c

```diff
@@ -0,0 +1,212 @@
+/**
+ * @file
+ *
+ * @author  Alex Kalmuk
+ * @date    05.05.2015
+ */
+
+#include <assert.h>
+#include <errno.h>
+#include <string.h>
+#include <util/math.h>
+
+#include <net/l2/ethernet.h>
+#include <net/netdevice.h>
+#include <net/inetdevice.h>
+#include <net/skbuff.h>
+#include <net/l0/net_entry.h>
+#include <drivers/usb/usb_driver.h>
+#include <drivers/usb/usb_dev_desc.h>
+#include <drivers/usb/usb_cdc.h>
+#include <mem/sysmalloc.h>
+#include <embox/unit.h>
+
+#include <kernel/time/timer.h>
+
+#include <kernel/printk.h>
+
+EMBOX_UNIT_INIT(usbnet_init);
+
+#define USBNET_TIMER_FREQ 1
+
+static struct usb_driver usbnet_driver;
+static const struct net_driver usbnet_drv_ops;
+
+static void usbnet_timer_handler(struct sys_timer *tmr, void *param);
+static void usbnet_rcv_notify(struct usb_request *req, void *arg);
+static int usbnet_init(void);
+static int usbnet_probe(struct usb_driver *drv, struct usb_dev *dev,
+		void **data);
+static void usbnet_disconnect(struct usb_dev *dev, void *data);
+static void usb_net_bulk_send(struct usb_dev *dev, struct sk_buff *skb);
+static int usbnet_xmit(struct net_device *dev, struct sk_buff *skb);
+
+struct usbnet_priv {
+	struct usb_dev *usbdev;
+	/* Storage for incoming packet */
+	char *data;
+	/* Current position in packet */
+	char *pdata;
+	/* Timer for polling */
+	struct sys_timer timer;
+};
+
+static int usbnet_init(void) {
+	return usb_driver_register(&usbnet_driver);
+}
+
+static void usbnet_rcv_notify(struct usb_request *req, void *arg) {
+	struct usb_endp *in_endp;
+	struct usb_dev *dev = req->endp->dev;
+	struct usb_class_cdc *cdc = usb2cdcdata(dev);
+	struct net_device *nic = (struct net_device *) cdc->privdata;
+	struct usbnet_priv *nic_priv = (struct usbnet_priv *) nic->priv;
+
+	in_endp = nic_priv->usbdev->endpoints[2];
+	assert(in_endp);
+
+	switch(req->req_stat) {
+	case USB_REQ_NOERR:
+		memcpy(nic_priv->pdata, req->buf, req->len);
+		nic_priv->pdata += req->len;
+		break;
+	default:
+		return;
+	}
+
+	/* End of packet */
+	if (req->len < in_endp->max_packet_size) {
+		struct sk_buff *skb;
+		size_t len;
+
+		len = nic_priv->pdata - nic_priv->data;
+		if ((skb = skb_alloc(len))) {
```

Comment:
IMO, it's better to stick to the common pattern:
```c
skb = skb_alloc(len);
if (!skb) {
    printk("usbnet: skbuff allocation failed\n");
} else {
    ...
}
```

I.e. deal with possible errors just after the allocation and forget about them right away.